docs/Guide.TypeScriptIntegrationJest.md

```diff
@@ -0,0 +1,111 @@
+### 0. Initialize Detox in your project
+
+Follow the [Getting Started](Introduction.GettingStarted.md) Guide to set up Detox with [Jest as your test runner](Guide.Jest.md) (optionally, set up Detox to [test Android](Introduction.Android.md) as well). When you are finished with that, we will make some changes to your project:
+
+- Remove Detox from `package.json` at the root level
+
+- Create a `package.json` in the `e2e` folder and add Detox and Jest to the project dependencies:
+
+```sh
+npm uninstall detox
+cd e2e
+npm init -y
+npm install -D detox jest
+```
+
+You should then move the settings for Detox from the root-level `package.json` to the `e2e/package.json`, but make sure that you update the `build` and `binaryPath` values for your Detox configurations:
+
+```json
+{
+  // ...
+  "detox": {
+    "configurations": {
+      "ios.sim.debug": {
+        "binaryPath": "../ios/build/Build/Products/Debug-iphonesimulator/YourAppName.app",
+        "build": "xcodebuild -project ../ios/YourAppName.xcodeproj -scheme YourAppName -configuration Debug -sdk iphonesimulator -derivedDataPath ../ios/build",
+        "type": "ios.simulator",
+        "name": "iPhone X"
+      },
+      "android.emu.debug": {
+        "binaryPath": "../android/app/build/outputs/apk/debug/app-debug.apk",
+        "build": "cd ../android; ./gradlew assembleDebug assembleAndroidTest -DtestBuildType=debug; cd ../e2e",
+        "type": "android.emulator",
+        "name": "Pixel_2_API_27"
+      },
+      "android.emu.release": {
+        "binaryPath": "../android/app/build/outputs/apk/release/app-release.apk",
+        "build": "cd ../android; ./gradlew assembleRelease assembleAndroidTest -DtestBuildType=release; cd ../e2e",
+        "type": "android.emulator",
+        "name": "Pixel_2_API_27"
+      }
+    }
+  }
+}
+```
+
+Before moving on, make sure that you can still successfully run your Detox tests.
+
+### 1. Add TypeScript + `ts-jest` to `e2e/package.json`
+
+```sh
+npm install --save-dev typescript ts-jest
+```
+
+### 2. Configure Jest to use `ts-jest`
+
+Modify your Jest configuration (`e2e/config.json` by default) to include the following properties
+
+```json
+{
+  "preset": "ts-jest",
+  "testEnvironment": "node",
+  "setupTestFrameworkScriptFile": "./init.ts"
+}
+```
+
+NB: this is mostly the same output of running `ts-jest config:init`, with `setupTestFrameworkScriptFile` being the only added property.
+
+### 3. `.js` -> `.ts`
+
+Convert all files in the `e2e` directory ending in `.js` to `.ts`
+
+### 4. Add typings for Detox, Jest, and Jasmine
+
+Add typings for Detox, Jest, and Jasmine (the latter two are used in `init.ts`), as well as for other modules that you use in your Detox tests.
+
+```sh
+npm install --save-dev @types/detox @types/jest @types/jasmine
+```
+
+You should now be able to run your Detox tests, written in TypeScript! If you're not writing your unit tests with TypeScript and Jest, you can skip the next section.
+
+### 5. Unit test collisions
```

Comment:
It be easier and more straightforward to just not export global functions with Detox on a TS project, and require them explicitly.

Reply:
Sorry, I'm not sure that I understand your comment? The issue mostly deals with the current typings for Detox since it declares `expect` as a global constant when Jest also does. For unit tests (that aren't utilizing Detox), `expect` should be defined as a `jest.Expect` whereas they should be `Detox.Expect` for e2e tests (otherwise, the TS compiler gets confused, per [this comment](https://github.com/wix/detox/issues/99#issuecomment-402206854))

Reply:
What if you setup Detox not to export Expect as as global?
https://github.com/wix/Detox/blob/master/docs/APIRef.DetoxObjectAPI.md#explicit-imports-during-initialization
Then you'll have to import everything manually like so:

```js
const {device, expect, element, by, waitFor} = require('detox');
```
This solves the collision issue, doesn't it?

Reply:
I hadn't realized that was possible before, but I would prefer to do that. Unfortunately, the current Detox typings are limiting us from doing that because `expect`, `device` et al. are declared as global constants, so you'll get a compiler error when you `require` them from Detox. For that snippet to work, we'd need to change the typings so that they aren't defined as global constants (which makes sense, because it's possible that they might not be globally defined); that would break some user's setups, however, because then they must declare those global constants in their code.

cc-ing @tareqelmasri and @stevechun as they are maintainers for the typings

Reply:
Another issue I noticed was that using `import ... from "detox"` will cause a compiler error, because `detox` isn't a module according to its typings file. `require`-ing it is fine because `require` returns `any`, which is okay as a work-around, but since I'm using TypeScript, I'd prefer to have those imports strongly typed, and I'd prefer to use import statements over `require` as that's what I normally use.

Reply:
I did, the error I got was that `File '~/myProject/node_modules/@types/detox/index.d.ts' is not a module.`, because Detox has only been defined as a namespace with some certain typings, and not also as a module that exports certain members. When I added the following, the error went away:

```ts
declare module detox {
  const detox: Detox.Detox;
  const device: Detox.Device;
  const expect: Detox.Expect;
  // Define the rest of the exports here.
  export {
    detox,
    device,
    expect,
    // rest of the exports
  ]
}
```

Do you mean for one definition file to support both styles? I think if possible, it won't be ergonomic; if we assume `expect` to be typed as `Detox.Expect | undefined`, then the compiler will complain if you try to use it: `Cannot invoke an object which is possibly 'undefined'.`. You'd need to explicitly cast it as `Detox.Expect` or sandwich `!` between `expect` and the opening parenthesis (e.g. `expect!()`) to tell the compiler that you know that it's defined. Also, whether those constants are defined globally can't be known until runtime, which is why I'm more preferential to leaving it up to users to define those global constants in their code, even if the default behavior of Detox is to define them globally.

Reply:
DefinitelyTyped/DefinitelyTyped#30410 has been merged, so I will update my PR and example projects to import Detox constants.